File: 6주차/LeetCode_124_Binary_Tree_Maximum_Path_Sum/dongjun.java
/**
 *  @문제접근
 *  후위순회로 트리를 순회하며 노드의 왼쪽, 오른쪽 서브트리의
 *  최대합을 구하고, 각각의 모든 합들 중 최대 값을 answer에 저장합니다.
 *
 *  @성능
 *  Runtime: 0 ms
 *  Memory Usage: 41.6 MB
 */
public class BinaryTreeMaximumPathSum {
    public static final int NULL = Integer.MIN_VALUE;
    public static int answer;
    public int maxPathSum(TreeNode root) {
        answer = Integer.MIN_VALUE;
        calcPathSum(root);
        return answer;
    }

    public int calcPathSum(TreeNode current){
        if(current == null)
            return NULL;

        int leftSubTreeSum = Math.max(0, calcPathSum(current.left));
        int rightSubTreeSum = Math.max(0, calcPathSum(current.right));
        int sumVal = current.val + leftSubTreeSum + rightSubTreeSum;

        answer = Math.max(answer, sumVal);

        return Math.max(leftSubTreeSum, rightSubTreeSum) + current.val;
    }
}

File: 10주차/LeetCode_94_Binary_Tree_Inorder_Traversal/dongjun.py
'''
1. 문제접근
    중위순회를 재귀적으로 구현해 풀었습니다.
    
2. 성능
    Runtime: 24ms
    Memory Usage: 14.1MB
    Time Complexity: O(N)
'''
class Solution:
    ret = []
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        self.ret = []
        
        return self.inorder(root)
    
    def inorder(self, node):
        if node is None:
            return self.ret
        
        self.inorder(node.left)
        self.ret.append(node.val)
        self.inorder(node.right)
        
        return self.ret
